Microsoft Entra ID provider

jjlakis opened the pull request:
## Microsoft Entra ID Provider
This PR contains a proposal for a new implementation of Azure provider, that:
* Is a child type of `OIDCPRovider`
* Supports group overage, multi-tenancy and V2 endpoint only - see https://github.com/oauth2-proxy/oauth2-proxy/issues/2384
* Prepares Azure provider to consume [client assertion token exchange](https://github.com/oauth2-proxy/oauth2-proxy/pull/2378) OIDC feature and thus support Azure Workload Identity.
* Has all Azure permissions and multi-tenancy considerations recognized and covered in docs.
* Replaces existing azure provider that is often misbehaving when it comes to permissions, groups and multi-tenancy.

### Changes
New OIDC-compliant provider is added - `ms-entra-id`. Provider-specific configuration flags are:

* `--ms-entra-id-allowed-tenant=<tenant-id>` - When specified, multi-tenant app authentication only allows tokens incomming from specific tenants.

Remarks:
* There's no flag for Tenant ID - only OIDC issuer URL is needed
* There's no groupGraphField support - RFC whether this is a valuable feature now or can be added separately.

### Usage
Detailed usage instructions with different scenarios, permissions, multi-tenancy considerations and examples with Azure Portal and Terraform are added to provider documentation.

### Features
Outside of the generic OIDC behavior, provider implements the following:
* **Group overage** - custom implementation of `EnrichSession()` checks for [group overage](https://learn.microsoft.com/en-us/security/zero-trust/develop/configure-tokens-group-claims-app-roles#group-overages) (If not disabled by `--ms-entra-id-skip-graph-groups`).
* **Allowed tenants for multi tenant apps** - Multi-tenant apps require `--insecure-oidc-skip-issuer-verification` flag to be set. Custom implementaion of `ValidateSession()` fills the security gap by comparing Tenant ID from incomming token with list specified in `--ms-entra-id-multi-tenant-allowed-tenant`.

### Other considerations
#### Scope 
This provider requires scope to be set explicitly (`openid` for azure only, and `openid profile email` for MS personal accounts). Default behavior [adds `groups` scope](https://github.com/oauth2-proxy/oauth2-proxy/blob/master/providers/oidc.go#L42-L44) when `--allowed-group` is set. This claim isn't supported by Azure (it's also [incompliant with OIDC](https://openid.net/specs/openid-connect-core-1_0.html#ScopeClaims)). Fix proposed separately: https://github.com/oauth2-proxy/oauth2-proxy/pull/2392

#### Missing e-mail in claim
When Azure user has no e-mail set (for example: it's created quickly for testing), oauth2-proxy fails with 500 with following:
```
[2024/01/14 20:32:23] [oauthproxy.go:853] Error creating session during OAuth2 callback: neither the id_token nor the profileURL set an email
```
When e-mail is added, everything works as expected. 

### How this was tested
This was carefully manually tested on Kubernetes with Traefik forward auth middleware. Following scenarios were covered:
* Single-tenant:
  * ✔ Minimal-permission app registration without groups claim enabled.
  * ✔ Minimal-permission app registration with groups claim enabled and with `--allowed-group`.
  * ✔ Group overage with `--allowed-group` + behavior for missing permissions.
* Multi-tenant:
  * ✔ Minimal-permission app registration without groups claim enabled, tested from 2 tenants & personal MS accounts
  * ✔ Minimal-permission app registration with groups claim enabled and with `--allowed-group`, tested from 2 tenants & personal MS account
  * ✔ Group overage with `--allowed-group`, tested from 2 tenants & personal MS account, check behavior for missing permissions.
  * ✔ ms-entra-id-multi-tenant-allowed-tenants - tested various combinations with 2 tenants & personal MS account
* Other tests
  * ✔ Docusaurus build

Implemented unit tests cover the corner cases - group overage & multi tenancy.

### Related PRs and issues
* https://github.com/oauth2-proxy/oauth2-proxy/pull/1972 - PR by @hskiba , updates existing provider to call Graph API only when needed. I like using Microsoft Go SDK to interact with Graph API, and would love to have something similar in this provider when we decide to support graph group field.
* https://github.com/oauth2-proxy/oauth2-proxy/pull/2378 - PR by @tuunit , adds client assertion auth to generic OIDC provider. These changes will allow to use Azure Workload Identity.
* https://github.com/oauth2-proxy/oauth2-proxy/issues/2384 - issue by @jjlakis , explains the differneces and outstandings in Azure vs. generic OIDC, describes background for this PR.

tuunit commented on docs/docs/configuration/providers/azure_entra_oidc.md:
```diff
@@ -0,0 +1,187 @@
+---
+id: azure_entra_oidc
+title: Azure Entra ID OIDC
```
Wouldn't the correct name be Microsoft Entry ID?
https://www.microsoft.com/en-us/security/business/identity-access/microsoft-entra-id

jjlakis replied:
Yes it is. Changed now in documentation

jjlakis replied:
Any preferences how to name the provider setting (currently `azure-oidc`). `ms-entra-id` could be an option.

hskiba replied:
I believe it's technically [Microsoft identity platform](https://learn.microsoft.com/en-us/entra/identity-platform/v2-overview) as it can authenticate more than just Entra users.

![about-microsoft-identity-platform](https://github.com/oauth2-proxy/oauth2-proxy/assets/16367391/1ac67fed-d6d1-47c8-b013-d1cba191ce7c)

jjlakis replied:
I think MS Entra ID name fits more into here, this is purely a successor of Azure AD, you can configure it fully from Azure through App registration as it was before.

tuunit commented on docs/static/img/azure-create-app-reg.gif:
I'm not the biggest fan of having gifs for documentation purposes. Especially because the encoding of the ones you provided is horrible, sorry 😅

I would rather prefer a image guide or the original screen capture embeded as a video:
```html
<video controls>
  <source src="screencapture.webm"/>
</video>
```

jjlakis replied:
You're right :smile: I added the raw MKV videos, does it look better?

tuunit commented on providers/azure_entra_oidc.go:
```diff
@@ -0,0 +1,166 @@
+)
+
+// NewAzureProvider initiates a new AzureProvider
+func NewAzureEntraOIDCProvider(p *ProviderData, opts options.Provider) *AzureEntraOIDCProvider {
```
How about?
```suggestion
func NewMicrosoftEntraIDProvider(p *ProviderData, opts options.Provider) *MicrosoftEntraIDProvider {
```

tuunit commented on pkg/apis/options/legacy_options.go:
```diff
@@ -488,6 +488,8 @@ type LegacyProvider struct {
 	AzureTenant                            string   `flag:"azure-tenant" cfg:"azure_tenant"`
 	AzureGraphGroupField                   string   `flag:"azure-graph-group-field" cfg:"azure_graph_group_field"`
+	AzureOIDCSkipGroupsFromGraph           bool     `flag:"azure-oidc-skip-groups-from-graph" cfg:"azure_oidc_skip_groups_from_graph"`
+	AzureOIDCMutliTenantAllowedTenants     []string `flag:"azure-oidc-multi-tenant-allowed-tenant" cfg:"azure_oidc_multi_tenant_allowed_tenants"`
```
This name seems a bit convoluted. Wouldn't `AzureAllowedTenants` be enough?

jjlakis replied:
Right, this variable is too long, I wanted to stick this with the flag name. Flag name is also a bit convoluted, but this setting applies only to the multi-tenant app registrations and I wanted to highlight it.

tuunit reviewed (commented):
Some remarks I already have before doing a indepth review.

jjlakis commented:
@tuunit I have renamed methods and settings to reflect Microsoft Entra ID name.

erik-h commented on pkg/apis/options/legacy_options.go:
```diff
@@ -488,6 +488,8 @@ type LegacyProvider struct {
 	AzureTenant                            string   `flag:"azure-tenant" cfg:"azure_tenant"`
 	AzureGraphGroupField                   string   `flag:"azure-graph-group-field" cfg:"azure_graph_group_field"`
+	MSEntraIDSkipGroupsFromGraph           bool     `flag:"ms-entra-id-skip-groups-from-graph" cfg:"ms_entra_id_skip_groups_from_graph"`
+	MSEntraIDMutliTenantAllowedTenants     []string `flag:"ms-entra-id-multi-tenant-allowed-tenant" cfg:"ms_entra_id_multi_tenant_allowed_tenants"`
```
Small typo: `MutliTenant` -> `MultiTenant`

kvanzuijlen replied:
```suggestion
	MSEntraIDMultiTenantAllowedTenants     []string `flag:"ms-entra-id-multi-tenant-allowed-tenant" cfg:"ms_entra_id_multi_tenant_allowed_tenants"`
```

kvanzuijlen commented on docs/docs/configuration/providers/index.md:
```diff
@@ -10,14 +10,14 @@ Valid providers are :
 
 - [Google](google.md) _default_
 - [Azure](azure.md)
+- [Microsoft Entra ID](ms_entra_id.md)
```
Please keep this list (other than Google, which is the default) sorted alphabetically

kvanzuijlen commented on docs/docs/configuration/providers/ms_entra_id.md:
```diff
@@ -0,0 +1,199 @@
+<details>
+    <summary>See Terraform example</summary>
+
+    resource "azuread_application" "auth" {
+        display_name     = "oauth2-proxy"
+        sign_in_audience = "AzureADMyOrg" # Others are also supported
+
+        web {
+            redirect_uris = [
+                "https://podinfo.lakis.tech/oauth2/callback",
+            ]
+        }
+
+        required_resource_access {
+            resource_app_id = "00000003-0000-0000-c000-000000000000" # Microsoft Graph
+            resource_access {
+                id   = "e1fe6dd8-ba31-4d61-89e7-88639da4683d" # User.Read delegated
+                type = "Scope"
+            }
+        }
+    }
+
+    resource "azuread_application_password" "apppass" {
+        application_id = azuread_application.auth.id
+    }
```
Does it make sense to put this in a code block? (\```...```)

jjlakis replied:
Right, by some reason docusaurus misbehaved before, with code block everything looks good
![image](https://github.com/oauth2-proxy/oauth2-proxy/assets/12859079/150e247c-ba58-4ff0-8032-a0aa3279c502)

kvanzuijlen commented on docs/docs/configuration/providers/ms_entra_id.md:
```diff
@@ -0,0 +1,199 @@
+<details>
+    <summary>See Terraform example</summary>
+
+    resource "azuread_application" "auth" {
+        display_name     = "oauth2-proxy"
+        sign_in_audience = "AzureADMyOrg" # Other alre also supported
+
+        group_membership_claims = [
+            "SecurityGroup"
+        ]
+        optional_claims {
+            id_token {
+                name = "groups"
+            }
+        }
+
+        web {
+            redirect_uris = [
+                "https://podinfo.lakis.tech/oauth2/callback",
+            ]
+        }
+
+        required_resource_access {
+            resource_app_id = "00000003-0000-0000-c000-000000000000" # Microsoft Graph
+            resource_access {
+                id   = "e1fe6dd8-ba31-4d61-89e7-88639da4683d" # User.Read delegated
+                type = "Scope"
+            }
+        }
+    }
+
+    resource "azuread_application_password" "apppass" {
+        application_id = azuread_application.auth.id
+    }
```
Same as above

kvanzuijlen commented on docs/docs/configuration/providers/ms_entra_id.md:
```diff
@@ -0,0 +1,199 @@
+<details>
+    <summary>See Terraform example</summary>
+
+    resource "azuread_application" "auth" {
+        display_name     = "oauth2-proxy"
+        sign_in_audience = "AzureADMyOrg" # Other alre also supported
+
+        group_membership_claims = [
+            "SecurityGroup"
+        ]
+        optional_claims {
+            id_token {
+                name = "groups"
+            }
+        }
+
+        web {
+            redirect_uris = [
+                "https://podinfo.lakis.tech/oauth2/callback",
+            ]
+        }
+
+        required_resource_access {
+            resource_app_id = "00000003-0000-0000-c000-000000000000" # Microsoft Graph
+            resource_access {
+                id   = "e1fe6dd8-ba31-4d61-89e7-88639da4683d" # User.Read delegated
+                type = "Scope"
+            }
+            
+            resource_access {
+                id   = "bc024368-1153-4739-b217-4326f2e966d0" # GroupMember.Read.All delegated - admin consent required
+                type = "Scope"
+            }
+
+        }
+    }
+
+    resource "azuread_application_password" "apppass" {
+        application_id = azuread_application.auth.id
+    }
```
Same as above

kvanzuijlen commented on docs/docs/configuration/providers/ms_entra_id.md:
````diff
@@ -0,0 +1,199 @@
+
+### Scope
+For Azure-only apps (multi-tenant and single-tenant), the only required oAuth scope is `openid`:
+```
````
```suggestion
```shell
```

kvanzuijlen commented on docs/docs/configuration/providers/ms_entra_id.md:
````diff
@@ -0,0 +1,199 @@
+- --scope=openid
+```
+For personal MS accounts, the scope has to be extended with `email` and `profile`:
+```
````
```suggestion
```shell
```

kvanzuijlen commented on docs/docs/configuration/providers/ms_entra_id.md:
````diff
@@ -0,0 +1,199 @@
+
+### Single-tenant
+Simple single-tenant configuration:
+```
````
```suggestion
```shell
```

kvanzuijlen commented on docs/docs/configuration/providers/ms_entra_id.md:
````diff
@@ -0,0 +1,199 @@
+
+### Multi-tenant
+Multi-tenant apps require to disable OIDC issuer verification, as `issuer` field in the [discovery document](https://login.microsoftonline.com/common/v2.0/.well-known/openid-configuration) is a template, not an exact value:
+```
````
```suggestion
```shell
```

kvanzuijlen commented on docs/docs/configuration/providers/ms_entra_id.md:
````diff
@@ -0,0 +1,199 @@
+```
+
+Configuration above insecurely allows all tenants, to allow specific tenants:
+```
````
```suggestion
```shell
```

kvanzuijlen commented on pkg/apis/options/legacy_options.go:
```diff
@@ -707,6 +711,11 @@ func (l *LegacyProvider) convert() (Providers, error) {
 
+	provider.MicrosoftEntraIDConfig = MicrosoftEntraIDOptions{
+		DisableGroupsFromGraph:    l.MSEntraIDSkipGroupsFromGraph,
+		MultiTenantAllowedTenants: l.MSEntraIDMutliTenantAllowedTenants,
```
```suggestion
		MultiTenantAllowedTenants: l.MSEntraIDMultiTenantAllowedTenants,
```

kvanzuijlen commented on docs/docs/configuration/providers/ms_entra_id.md:
```diff
@@ -0,0 +1,199 @@
+id: ms_entra_id
+title: Microsoft Entra ID
+---
+OIDC-compliant provider for Microsoft Entra ID (Azure AD successor) application registrations. This providers support Azure oAuth V2 endpoint only, if you want to use V1 endpoint, try legacy `azure` provider. 
```
```suggestion
OIDC-compliant provider for Microsoft Entra ID (Azure AD successor) application registrations. This provider supports the Azure OAuth V2 endpoint only, if you want to use the V1 endpoint, try the legacy [azure](azure.md) provider.

kvanzuijlen commented on docs/docs/configuration/providers/ms_entra_id.md:
```diff
@@ -0,0 +1,199 @@
+OIDC-compliant provider for Microsoft Entra ID (Azure AD successor) application registrations. This providers support Azure oAuth V2 endpoint only, if you want to use V1 endpoint, try legacy `azure` provider. 
+
+## Configure application registration
+To start, create an App registration with minimal permissions, assign redirect URI and generate secret. All account types are supported (Single tenant, multi tenant, multi tenant with MS accounts, MS accounts only).
```
```suggestion
To start, create an App registration with minimal permissions, assign a redirect URI, and generate a secret. All account types are supported (Single tenant, multi-tenant, multi-tenant with MS accounts, and MS accounts only).

kvanzuijlen commented on docs/docs/configuration/providers/ms_entra_id.md:
```diff
@@ -0,0 +1,199 @@
+This is sufficient for a simple authentication scenario.
+
+### Configure `groups` claim
+If you want to make use of groups (for example, use `--allowed-group` feature of oauth2-proxy or authorize based on groups inside your service), you need to configure `groups` claim to be present in the ID token:
```
```suggestion
If you want to make use of groups (i.e. `--allowed-group` or authorize based on groups inside your service), you need to configure the `groups` claim to be present in the ID token:

kvanzuijlen commented on docs/docs/configuration/providers/ms_entra_id.md:
```diff
@@ -0,0 +1,199 @@
+
+
+### Configure group overage support
+Azure has a limit of 200 groups in the JWT. If you can't avoid such a bug number and still want to access the groups, you need to grant `GroupMember.Read.All` delegated permission to the app registration so oauth2-proxy can read all the groups from Graph API. **NOTE**: This permission by default requires an admin consent!
```
```suggestion
Azure has a limit of 200 groups in the JWT. If you can't avoid such a big number and still want to access the groups, you need to grant `GroupMember.Read.All` delegated permission to the app registration so oauth2-proxy can read all the groups from Graph API. **NOTE**: This permission by default requires admin consent!

kvanzuijlen commented on docs/docs/configuration/providers/ms_entra_id.md:
```diff
@@ -0,0 +1,199 @@
+</details>
+
+## Configure provider
+Provider is OIDC-compliant, so all the OIDC parameters are honoured. Additional provider-specific configuration parameters are:
+* `ms-entra-id-skip-groups-from-graph` - never read groups from Graph API, even when ID token indicates that there's a group overage. Set if you expect group overage in some cases, but still don't want to assign wide `GroupMember.Read.All`. Defaults to `false`. If you don't need groups, consider skipping `groups` claim in the app registration.
+* `ms-entra-id-multi-tenant-allowed-tenant` - speciy list of allowed tenants to be authenticated through multi-tenant app. When not set, all tenants are allowed. Defaults to `[]` (all tenants).
```
```suggestion
The provider is OIDC-compliant, so all the OIDC parameters are honored. Additional provider-specific configuration parameters are:
* `ms-entra-id-skip-groups-from-graph` - never read groups from Graph API, even when the ID token indicates that there's a group overage. Set if you expect group overage in some cases, but still don't want to assign wide `GroupMember.Read.All`. Defaults to `false`. If you don't need groups, consider skipping the `groups` claim in the app registration.
* `ms-entra-id-multi-tenant-allowed-tenant` - specify a list of allowed tenants to be authenticated through a multi-tenant app. When not set, all tenants are allowed. Defaults to `[]` (all tenants).
```

kvanzuijlen commented on docs/docs/configuration/providers/ms_entra_id.md:
```diff
@@ -0,0 +1,199 @@
+* `ms-entra-id-multi-tenant-allowed-tenant` - speciy list of allowed tenants to be authenticated through multi-tenant app. When not set, all tenants are allowed. Defaults to `[]` (all tenants).
+
+### Scope
+For Azure-only apps (multi-tenant and single-tenant), the only required oAuth scope is `openid`:
```
```suggestion
For Azure-only apps (multi-tenant and single-tenant), the only required OAuth scope is `openid`:
```

kvanzuijlen commented on docs/docs/configuration/providers/ms_entra_id.md:
````diff
@@ -0,0 +1,199 @@
+```
+- --scope=openid profile email
+```
+It's recommended to configure the scopes explicitly, otherwise you may experience issues with allowing groups (Azure doesn't support `groups` scope which is automatically included when you configure allowed groups).
````
```suggestion
It's recommended to configure the scopes explicitly, otherwise, you may experience issues with allowing groups (Azure doesn't support the `groups` scope, which is automatically included when you configure allowed groups).
```

kvanzuijlen commented on docs/docs/configuration/providers/ms_entra_id.md:
````diff
@@ -0,0 +1,199 @@
+```
+
+### Multi-tenant
+Multi-tenant apps require to disable OIDC issuer verification, as `issuer` field in the [discovery document](https://login.microsoftonline.com/common/v2.0/.well-known/openid-configuration) is a template, not an exact value:
````
```suggestion
Multi-tenant apps require you to disable OIDC issuer verification, as `issuer` field in the [discovery document](https://login.microsoftonline.com/common/v2.0/.well-known/openid-configuration) is a template, not an exact value:

kvanzuijlen commented on docs/docs/configuration/providers/ms_entra_id.md:
````diff
@@ -0,0 +1,199 @@
+- --scope=openid profile email
+```
+
+Configuration above insecurely allows all tenants, to allow specific tenants:
````
```suggestion
The configuration above insecurely allows all tenants. To allow specific tenants, use the configuration below as an example:

kvanzuijlen reviewed (dismissed):
I did a quick read-through of the documentation. @tuunit can you take a look at the Golang code?

Also, are we 100% sure the Azure AD code can be deleted immediately? As this would be a breaking change I think I'd slightly prefer the removal in a separate PR.

jjlakis commented:
I have revisited & rebased. I've addressed documentation comments and tested with docusaurus. Documentation look good:
<details>

![image](https://github.com/oauth2-proxy/oauth2-proxy/assets/12859079/4e2ae12a-1283-46f0-bbfc-f0930da00b5c)

</details>

I have also run so simple testes to ensure everything still works :ok_hand:

JoelSpeed commented:
In general, +1 to the direction this is going, will try to find time for a thorough review. Thank you for the in-depth analysis you've done there.

tuunit commented on docs/static/azure-videos/create-app-reg.mkv:
i've uploaded the files to youtube

https://youtube.com/watch?v=IUNfxhOzr4E

tuunit commented on docs/static/azure-videos/create-groups-claim.mkv:
https://youtu.be/QZmP5MKEJis

tuunit commented on docs/static/azure-videos/group-overage.mkv:
https://youtu.be/6pNsU7TY1AU

tuunit commented on pkg/apis/options/legacy_options.go:
```diff
@@ -488,6 +488,8 @@ type LegacyProvider struct {
 	KeycloakGroups                         []string `flag:"keycloak-group" cfg:"keycloak_groups"`
 	AzureTenant                            string   `flag:"azure-tenant" cfg:"azure_tenant"`
 	AzureGraphGroupField                   string   `flag:"azure-graph-group-field" cfg:"azure_graph_group_field"`
+	MSEntraIDSkipGroupsFromGraph           bool     `flag:"ms-entra-id-skip-groups-from-graph" cfg:"ms_entra_id_skip_groups_from_graph"`
```
I would just call them entra without the MS prefix
```suggestion
	EntraIDSkipGroupsFromGraph           bool     `flag:"entra-id-skip-groups-from-graph" cfg:"ms_entra_id_skip_groups_from_graph"`
```

tuunit commented on pkg/apis/options/legacy_options.go:
```diff
@@ -488,6 +488,8 @@ type LegacyProvider struct {
 	AzureTenant                            string   `flag:"azure-tenant" cfg:"azure_tenant"`
 	AzureGraphGroupField                   string   `flag:"azure-graph-group-field" cfg:"azure_graph_group_field"`
+	MSEntraIDSkipGroupsFromGraph           bool     `flag:"ms-entra-id-skip-groups-from-graph" cfg:"ms_entra_id_skip_groups_from_graph"`
+	MSEntraIDMultiTenantAllowedTenants     []string `flag:"ms-entra-id-multi-tenant-allowed-tenant" cfg:"ms_entra_id_multi_tenant_allowed_tenants"`
```
Instead of adding the booling at the bottom cannot we just check if the list is empty or not?

or is the list optional?

```suggestion
	EntraIDAllowedTenants     []string `flag:"entra-id-allowed-tenant" cfg:"ms_entra_id_multi_tenant_allowed_tenants"`
```

jjlakis replied:
List is optional, by default accept all tenants when using multi-tenant apps

tuunit commented on pkg/apis/options/legacy_options.go:
```diff
@@ -550,6 +552,8 @@ func legacyProviderFlagSet() *pflag.FlagSet {
 	flagSet.String("azure-tenant", "common", "go to a tenant-specific or common (tenant-independent) endpoint.")
 	flagSet.String("azure-graph-group-field", "", "configures the group field to be used when building the groups list(`id` or `displayName`. Default is `id`) from Microsoft Graph(available only for v2.0 oidc url). Based on this value, the `allowed-group` config values should be adjusted accordingly. If using `id` as group field, `allowed-group` should contains groups IDs, if using `displayName` as  ...
+	flagSet.Bool("ms-entra-id-skip-groups-from-graph", false, "when set, MS Entra ID provider doesn' try to read groups from microsoft graph even in case of group overage")
+	flagSet.StringSlice("ms-entra-id-multi-tenant-allowed-tenant", []string{}, "list of tenants allowed for MS Entra ID multi-tenant application")
```
duplicate?

jjlakis replied:
You mean duplicate with what exactly?

tuunit commented on pkg/apis/options/providers.go:
```diff
@@ -30,6 +30,8 @@ type Provider struct {
 	KeycloakConfig KeycloakOptions `json:"keycloakConfig,omitempty"`
 	// AzureConfig holds all configurations for Azure provider.
 	AzureConfig AzureOptions `json:"azureConfig,omitempty"`
+	// AzureConfig holds all configurations for Azure provider.
```
```suggestion
	// MicrosoftEntraIDConfig holds all configurations for the Entra ID provider.
```

tuunit commented on pkg/apis/options/providers.go:
```diff
@@ -101,6 +103,9 @@ const (
 	// AzureProvider is the provider type for Azure
 	AzureProvider ProviderType = "azure"
 
+	// AzureProvider is the provider type for Azure Entra OIDC
```
```suggestion
	// MicrosoftEntraIDProvider is the provider type for Entra OIDC
```

tuunit commented on pkg/apis/options/providers.go:
```diff
@@ -155,6 +160,11 @@ type AzureOptions struct {
 
+type MicrosoftEntraIDOptions struct {
+	DisableGroupsFromGraph    bool     `json:"disableGroupsFromGraph,omitempty"`
+	MultiTenantAllowedTenants []string `json:"multiTenantAllowedTenants,omitempty"`
```
how about?

```suggestion
	AllowedMultiTenants []string `json:"multiTenantAllowedTenants,omitempty"`
```

tuunit commented on providers/ms_entra_id.go:
```diff
@@ -0,0 +1,166 @@
+	multiTenantAllowedTenants []string
+
+	microsoftGraphURL *url.URL
+	isMultiTenant     bool
```
Again is this boolean necessary because the list of allowed tenants is optional?

jjlakis replied:
I have run some tests and this is actually indeed redundant and I will remove it.
List of allowed tenants only makes sense for multi-tenants app, however it can serve as an redundant check for single-tenant apps too. I have also simplified the flag names.

tuunit commented on providers/ms_entra_id.go:
```diff
@@ -0,0 +1,166 @@
+	}
+)
+
+// NewAzureProvider initiates a new MicrosoftEntraIDProvider
```
please fix description

tuunit commented on providers/ms_entra_id.go:
```diff
@@ -0,0 +1,166 @@
+			return false
+		}
+
+		if p.multiTenantAllowedTenants != nil {
```
```suggestion
		if len(p.multiTenantAllowedTenants) > 0 {
```

tuunit commented on providers/ms_entra_id.go:
```diff
@@ -0,0 +1,166 @@
+	var claimNamesMap map[string]string
+
+	if claimNames != nil {
+		claimNamesMap = cast.ToStringMapString(claimNames)
```
is it necessary to use an external dependency for this remapping?

jjlakis replied:
What we check here is a following claim:
```
"_claim_names": {
  "groups": "src1"
}
```
So I need to somehow transform content of a claim it to `map[string]string`. I must have found this library in [claim_extractor.go](https://github.com/oauth2-proxy/oauth2-proxy/blob/master/pkg/providers/util/claim_extractor.go#L15). Any better method appreciated :smile:

tuunit commented on providers/ms_entra_id.go:
```diff
@@ -0,0 +1,166 @@
+	return p.OIDCProvider.ValidateSession(ctx, session)
+}
+
+// checkGroupOverage checks is claims present in ID Token indicate group overage
```
```suggestion
// checkGroupOverage checks if the claims for the group membership are present in the ID Token
// https://learn.microsoft.com/en-us/security/zero-trust/develop/configure-tokens-group-claims-app-roles#group-overages
```

jjlakis replied:
This is not exactly checking the group membership, but will fix.

tuunit commented on providers/ms_entra_id.go:
```diff
@@ -0,0 +1,166 @@
+
+	groupsURL := fmt.Sprintf("%s/transitiveMemberOf?$select=id", p.microsoftGraphURL)
+
+	jsonRequest, err := requests.New(groupsURL).WithContext(ctx).WithHeaders(groupsHeaders).Do().UnmarshalSimpleJSON()
```
Please format into a multiline statement for better readability

tuunit commented on providers/ms_entra_id.go:
```diff
@@ -0,0 +1,166 @@
+	}
+
+	var groups []string
+	for i := range jsonRequest.Get("value").MustArray() {
```
what happens if value is empty?

jjlakis commented:
@tuunit I have rebased and addressed most of your comments :ok_hand:

tuunit commented:
> @tuunit I have rebased and addressed most of your comments 👌

Amazing I'll try to do a review this week!

tuunit commented:
@jjlakis I've created an Azure test account and I'll do some testing tomorrow and on Tuesday. I hope that we can get this merged soon :)

rbange commented:
Still relevant

tuunit commented:
For those who are interested. OAuth2-Proxy got an official sponsoring from Azure and we now have a large account to do fully integrated and automated testing. Unfortunately, it will still take some time to get the testing setup working and the PR finalized.

jkroepke commented:
Would it be possible to keep client certificates in mind? 

Microsoft doesn't recommend plain password credentials

> Consider the following guidance related to certificates and secrets:
> 
> Always use [certificate credentials](https://learn.microsoft.com/en-us/entra/identity-platform/certificate-credentials) whenever possible and don't use password credentials, also known as secrets. While it's convenient to use password secrets as a credential, when possible use x509 certificates as the only credential type for getting tokens for an application.

Ref: https://learn.microsoft.com/en-us/entra/identity-platform/security-best-practices-for-app-registration#certificates-and-secrets

jjlakis commented:
@jkroepke This provider is pretty much a pre-work for consuming the certificate credentials instead of secrets (see the PR description). Using certificate credentials will be rolled out more generically (as part of OIDC feature) and then support via Entra.

tuunit commented on providers/ms_entra_id.go:
```diff
@@ -0,0 +1,162 @@
+}
+
+// EnrichSession checks for group overage and passes through to generic EnrichSession()
+func (p *MicrosoftEntraIDProvider) EnrichSession(ctx context.Context, session *sessions.SessionState) error {
+	err := p.OIDCProvider.EnrichSession(ctx, session)
+
```
Regarding issuer verification:

`EnrichSession` is always called no matter if it's the first authentication or following reqeuests.

Therefore we can add a verification in here to check against every tenant id provided by the config. We will need to parse the JWT token in `session.IDToken` can be as simple as:

```go
	rawPayload, _ := base64.StdEncoding.DecodeString(strings.Split(session.IDToken, ".")[1])

	payload := make(map[string]interface{})
	json.Unmarshal([]byte(rawPayload), &payload)

	issuer := payload["iss"].(string)
```

https://go.dev/play/p/TAi9QePsJ7_L

jjlakis replied:
The concern I've mentioned on the call is [OpenID Provider Configuration Validation](https://openid.net/specs/openid-connect-discovery-1_0.html#ProviderConfigurationValidation)  that (I believe) happens here :fist_right: [provider.go#L63
](https://github.com/oauth2-proxy/oauth2-proxy/blob/master/pkg/providers/oidc/provider.go#L63) and happens on the provider initialization. It's disabled by the same flag that skips validation of the ID token which is a completely separate process. I will include the detailed information on the documentation.

Currently check against the Issuer URL is implemented in `ValidateSession()`, you think this should be moved to `EnrichSession()`?

tuunit replied:
You are right and for multi tenancy we have to disable odic verification because in that case the common tenant url is used but the resulting oidc token will contain a specific issuer which we can then check inside the enrich session method. it will act as a second layer of verification.

tuunit replied:
and indeed i got so into the generic implementation that i totally missed that you already implemented my idea using the proper `ValidateSession` flow.

tuunit replied:
I'll go through everything once more tomorrow but I think we are ready to merge 🎉

jjlakis commented:
**Asking for an ultimate review!** 

I have revisited, rebased and slightly refactored the code. The notable change is: Turned out no admin consent is required in any case of group overage, the least privilege scope is `User.Read.` I made this clear in the documentation. Also, when groups are unable to retrieve, there's no 500 error, but 403. As this check became soft, I removed the `skip-groups-from-graph` flag.

I have tested all the scenarios with 3 different app registrations according to following diagram:
![entra-e2e-all-cases](https://github.com/user-attachments/assets/4d939ed8-c522-41a4-ac49-622a14de331c)

This will be also used as a baseline for upcomming e2e testing framework.

tuunit commented:
> **Asking for an ultimate review!**
> ...
> This will be also used as a baseline for upcomming e2e testing framework.

I'll add some final tweaks tomorrow and after that. It is ready for one final review by Joel.

tuunit reviewed (dismissed):
LGTM

Awaiting final review by @JoelSpeed 

E2E Test integration will be added in two further PRs.

One PR about general restructuring for e2e integration with Providers we can test "locally" inside the workflow like Dex, Keycloak and Gitlab and an additional PR regarding the e2e integration with Azure.

tuunit commented on docs/docs/configuration/alpha_config.md:
```diff
@@ -385,6 +385,16 @@ character.
+
+| Field | Type | Description |
+| ----- | ---- | ----------- |
+| `allowedTenants` | _[]string_ |  |
```
Autogenerated docs missing a description because the godocs are missing!

tuunit commented on docs/docs/configuration/providers/azure_ad.md:
I added a big note to the actual azure provider. This document was just about how to configure it for Azure Active Directory. Which doesn't exist anymore because it was replaced by Entra ID. Their is no reason to keep it for a none existing service.

tuunit commented on docs/docs/configuration/providers/ms_entra_id.md:
```diff
@@ -0,0 +1,154 @@
+client_id="<client-id>"
+client_secret="<client-secret>"
+scope="openid"
+allowed_groups=["ac51800c-2679-4ecb-8130-636380a3b491"]
```
Yes as per every provider

JoelSpeed replied:
In the future, maybe it would be possible/better to have the group names/email addresses? But I guess that doesn't make things unique at the provider side 🤔

tuunit commented on docs/docs/configuration/providers/ms_entra_id.md:
````diff
@@ -0,0 +1,154 @@
+### Example configurations
+Single-tenant app without groups (*groups claim* not enabled). Consider using generic OIDC provider:
+```shell
+provider="entra-id"
````
What exactly do you mean?

tuunit replied:
We don't really have a naming scheme. Most are just lower case like `github`, `bitbucket`, `google` or `bitbucket` but we have `login.gov` as well.

JoelSpeed replied:
Perhaps at some point we should standardise these, so they are all, for example, PascalCase or something?

tuunit commented on providers/ms_entra_id.go:
```diff
@@ -0,0 +1,195 @@
+	groupsHeaders := makeAuthorizationHeader(tokenTypeBearer, s.AccessToken, nil)
+	groupsHeaders.Add("ConsistencyLevel", "eventual")
+
+	groupsURL := fmt.Sprintf("%s/transitiveMemberOf?$select=id&$top=999", p.microsoftGraphURL)
```
https://learn.microsoft.com/en-us/graph/api/orgcontact-list-transitivememberof?view=graph-rest-1.0&tabs=http#optional-query-parameters

The default and maximum page sizes are 100 and 999 objects respectively.

tuunit replied:
@jjlakis do we need to introduce pagination here? Because there could be more than 999 groups.

For proper pagination support. We would need to do one call using `"%s/transitiveMemberOf?$count` and then do multiple calls like so:

```go
for skip = 0; skip < count; skip = skip+100 {
  call("%s/transitiveMemberOf?$skip=%d&$top=%d", skip, skip+100)
}
```

tuunit replied:
Found documentation on how paging works. The skip parameter works but the following is recommend:
https://learn.microsoft.com/en-us/graph/paging?view=graph-rest-1.0&tabs=http

tuunit replied:
Already implemented a solution based on the docs. Adding two test right now.

tuunit replied:
Paging is now added including unit testing

tuunit reviewed (commented):
@JoelSpeed applied your review suggestions

htpawel commented:
Hi! Any update on this? Do we have any ETA?

JoelSpeed commented on pkg/apis/options/providers.go:
```diff
@@ -155,6 +160,14 @@ type AzureOptions struct {
 }
 
+type MicrosoftEntraIDOptions struct {
+	// List of allowed tenants. In case of multi-tenant apps, incoming tokens are
```
Nit, should start with the field name, like other options
```
AllowedTenants is a list of allowed Entra ID tenants. In case of...
```

JoelSpeed commented on providers/ms_entra_id.go:
```diff
@@ -0,0 +1,188 @@
+			return fmt.Errorf("unable to enrich session: %v", err)
+		}
+	} else {
+		logger.Printf("entra group overage not found")
```
Probably don't need this else clause, might pollute logs a bit?

jjlakis replied:
Will remove this one.

JoelSpeed commented on providers/ms_entra_id.go:
```diff
@@ -0,0 +1,188 @@
+	}
+
+	if hasGroupOverage {
+		logger.Printf("entra overage found, reading groups from Graph API")
```
Do we really want to log this?

jjlakis replied:
As this is an exceptional case, I vote for keeping it.

d-led commented:
tried it out. Works as desired. Nice work!

Yc-Chen commented:
Hi, just mentioning my issue here, hoping to get some help from Microsoft people: https://github.com/oauth2-proxy/oauth2-proxy/issues/2984. I still fail with group overage even when I got the admin consent.
